### ISSUE TITLE
DNS leak in Linux/Wireguard when LAN gateway/DNS is 172.16.0.1

### ISSUE THREAD

**Issue by im-batman** [trailofbits/algo#1422]
**Describe the bug**

**(Update 5/6/2019)** Back to point A. Addressing conflicts between physical network and VPS-hosted DNS server indeed create DNS leaks. Even after removing this addressing conflict, DNS continues to leak secondary lookups over the physical interface upon receipt of NXDOMAIN responses. I'm observing these lookups after NXDOMAIN responses to blacklisted domains.

5/5/2019 update is moot as I was misconfiguring based on Ubuntu Server instructions that don't apply to Ubuntu Desktop because of NetworkManager conflicts.

~~**(Update 5/5/2019)** I examined this further and found that the route is not being created to `172.16.0.0/24` regardless of the LAN addressing scheme. This results in DNS leaks on all Ubuntu/Wireguard clients since the host cannot reach the VPN's DNS on `172.16.0.1`.~~

When using an Algo-provisioned VPN with the Wireguard client in Ubuntu 19.04, if the client's home network is `172.16.0.0/24` and the gateway and DNS are both `172.16.0.1` (weird edge case but it just happened to me), this will result in a DNS leak. The DNS requests will continue to flow through the LAN `172.16.0.1` because the VPN route was not established successfully for that subnet (it already existed).

The issue is a collision between Algo's hardcoded addressing scheme and the addressing scheme of my home network.

**To Reproduce**

1. Roll out a VPN service to any provider (I used Azure), default settings plus DNS ad blocker.
2. Configure an Ubuntu 19.04 client with an IP address of 172.16.0.101/24 and a gateway/DNS of 172.16.0.1.
3. Install and configure Wireguard as specified in algo/docs/client-linux-wireguard.md.
4. Start wireguard service via systemd.
5. Check https://www.whoer.net to view which DNS is detected.

**Expected behavior**

Using https://www.whoer.net to verify external IP and DNS, I expected to see N/A for my DNS. Instead I saw my ISP's DNS, indicating a leak.

**Additional context**

No additional context.

**Full log** (note changes to local addressing 5/5/2019)

```bash
user@user:~$ netstat -nr
Kernel IP routing table
Destination     Gateway         Genmask         Flags   MSS Window  irtt Iface
0.0.0.0         172.16.7.1      0.0.0.0         UG        0 0          0 enp0s31f6
169.254.0.0     0.0.0.0         255.255.0.0     U         0 0          0 enp0s31f6
172.16.7.0      0.0.0.0         255.255.255.0   U         0 0          0 enp0s31f6

user@user:~$ sudo systemctl start wg-quick@wg0

user@user:~$ netstat -nr
Kernel IP routing table
Destination     Gateway         Genmask         Flags   MSS Window  irtt Iface
0.0.0.0         172.16.7.1      0.0.0.0         UG        0 0          0 enp0s31f6
10.19.49.0      0.0.0.0         255.255.255.0   U         0 0          0 wg0
169.254.0.0     0.0.0.0         255.255.0.0     U         0 0          0 enp0s31f6
172.16.7.0      0.0.0.0         255.255.255.0   U         0 0          0 enp0s31f6
```

**Comment by davidemyers** [comment]
I don't think there's really a bug here, just an unfortunate network address overlap. WireGuard's `wg-quick` intends to leave traffic destined to your local network alone, and Algo's default DNS server address looks to your client like it's on your local network.

The workaround is to change `local_service_ip: 172.16.0.1` in `config.cfg` before you deploy.

**Comment by im-batman** [comment]
I hear what you're saying and I agree that it's an unfortunate overlap, and maybe it isn't a "bug" as such, but it is a problem that I believe deserves attention. A few thoughts: 
1. DNS leaks are the most insidious vulnerability of a VPN connection. If you're leaking DNS you might as well not be using a VPN at all. Impact: **critical**.
2. DNS leaks are very silent. There's no error or indication that DNS isn't working unless you deliberately check and know what you're looking for. If you provision your VPN on a 192.x network (where everything checks out and DNS is working great) and then migrate over to a 172.16.0.0/24 network, you might forget to check that your DNS is still going where it should be. If you don't, you're blown. Likelihood that this will eventually happen in the wild: **very high**.
3. The defeat for this vulnerability is extremely low-tech. As an attacker I need only set up wifi honeypots with 172.16.0.0/24 addressing, and then I can monitor and MITM DNS from every Ubuntu/Algo user who connects even while they're "protected." Ease of targeted exploit: **trivial**.

Bottom line, I agree that some of this is up to the user, but this scenario feels like a subtle "gotcha" that breaks the product too easily, completely, and quietly.

**Comment by davidemyers** [comment]
I think you make some good points. Note that the WireGuard clients for iOS and macOS also implement "local LAN bypass" behavior and I wouldn't be surprised if the same was true of Android, so this issue possibly applies to many WireGuard users (assuming none of these WireGuard implementations configure the client to force traffic to the configured DNS servers to only go over the tunnel).

I'm not sure if there's anything Algo can do about this, though. Perhaps someone will have some ideas.

**Comment by jackivanov** [comment]
It would be good if we can cover this case in the docs

**Comment by davidemyers** [comment]
@jackivanov Would changing the default DNS address be worth it to reduce the chances of an accidental conflict? Like maybe `10.19.50.1` (I haven't tested this)?

If an Algo client ends up on a network that uses `10.19.0.0/16` more than DNS will probably break anyway.

**Comment by im-batman** [comment]
Actually this is happening to me on other private ranges as well. It appears as though the route to 172.16.0.0/24 via wg0 isn't being created at all, so DNS look-ups aren't making it to 172.16.0.1 as intended.

This would logically result in leaks on **all** Ubuntu/Wireguard clients, not just those on specific subnets. I believe this *is* actually a bug and it is considerably larger than originally though.

**Comment by davidemyers** [comment]
The bug could be in the instructions.

Are you using `resolvectl` or `openresolv`? If `resolvectl`, what do you see for your WireGuard interface when you run `resolvectl`? I see:

```
Link 3 (wg0)
      Current Scopes: DNS
DefaultRoute setting: yes
       LLMNR setting: yes
MulticastDNS setting: no
  DNSOverTLS setting: no
      DNSSEC setting: no
    DNSSEC supported: no
  Current DNS Server: 172.16.0.1
         DNS Servers: 172.16.0.1
          DNS Domain: ~.
```

**Comment by im-batman** [comment]
I'm using `resolvectl` and it returns results indentical to yours.

I ran `tcpdump` while accessing https://whoer.net and found that DNS requests went out over both the `wlan0` and `wg0` interfaces in more or less equal quantities. It's not clear yet why some went over the VPN tunnel and some didn't; the only distinction I can find is that the IPv6 (AAAA) lookups all went over the tunnel, whereas the native wlan interface saw only IPv4 (A) lookups. Additionally whoer.net showed my ISP's DNS address, once again confirming leaks.

**Comment by davidemyers** [comment]
What does `resolvectl` show you for `wlan0`?

**Comment by im-batman** [comment]
```
Link 5 (wg0)
      Current Scopes: DNS
DefaultRoute setting: yes
       LLMNR setting: yes
MulticastDNS setting: no
  DNSOverTLS setting: no
      DNSSEC setting: no
    DNSSEC supported: no
  Current DNS Server: 172.16.0.1
         DNS Servers: 172.16.0.1
          DNS Domain: ~.

Link 3 (wlan0)
      Current Scopes: DNS
DefaultRoute setting: yes
       LLMNR setting: yes
MulticastDNS setting: no
  DNSOverTLS setting: no
      DNSSEC setting: no
    DNSSEC supported: no
  Current DNS Server: 172.16.7.1
         DNS Servers: 172.16.7.1
                      my.isp.my.isp
                      my.isp.my.isp
          DNS Domain: ~.
                      my.domain

Link 2 (eth0)
      Current Scopes: none
DefaultRoute setting: no
       LLMNR setting: yes
MulticastDNS setting: no
  DNSOverTLS setting: no
      DNSSEC setting: no
    DNSSEC supported: no
```

**Comment by im-batman** [comment]
I think I'm looking the problem right in the eye... `wlan0` and `wg0` are both DefaultRoute for DNS lookups. So the question is, how do we make it so users can elect to automatically turn off DefaultRoute for their wlan or eth interfaces when they connect?

**Comment by davidemyers** [comment]
I'm concerned about the `~.` in the settings for `wlan0`.

Are you using Ubuntu Desktop (the instructions are for Ubuntu Server)? So NetworkManager is being used?

**Comment by im-batman** [comment]
... oof. I am indeed using Desktop and did not realize these were for Server only. Yep, that changes things. Network manager is being used to to establish wlan connection but not Wireguard.

**Comment by davidemyers** [comment]
I've heard that recent versions of NetworkManager can import WireGuard configs but I haven't tried it (something about using `nmcli` -- I'm not very familiar with NetworkManager). If you can make it work, note how you did it here and I'll update the documentation.

**Comment by tidelake** [comment]
I'm puzzled that no attention goes to what the client might have going on. Can you list some resources for client setup?

**Comment by im-batman** [comment]
I haven't attempted yet to import Wireguard configs into NetworkManager but I did some additional testing. I removed NetworkManager altogether and tested the `172.16.0.0/24` network again. DNS leaks, backing my original hypothesis that the conflicting routes are actually a problem for both Ubuntu desktop and likely Ubuntu server as well. `resolvectl` indicated that only `wg0` was set as DNS DefaultRoute.

Then I built a new Algo server stack with a unique address for my internal DNS (`172.x.y.1`).

Some new and interesting observations from that experiment:

1. Most DNS lookups now go over the Wireguard tunnel but **some still fail over to the physical interface**. 
2. Fortunately there is now a common thread. I noted from the `tcpdump` output that all of the requests that went over the physical interface were appended with my domain suffix and they were lookups to ad domains. I'm deducing that these lookups are being made after receiving NXDOMAIN responses from the adblock service. Nonetheless, on an uncontrolled physical network this environment leaks lookups. This will continue to happen as long as the client keeps its local network routes prioritized over the VPN route.

I'm not sure but it might help reduce these instances if the adblock server returns a `127.0.0.1` A-record instead of NXDOMAIN.

**Comment by im-batman** [comment]
I would also make the general recommendation that `local_service_ip` be a randomized private address on a very small subnet, with an option to modify it when running the `algo` wizard script.

**Comment by jackivanov** [comment]
I'm thinking whether we can prevent the conflicts by using only IPv6 for DNS.

@im-batman Could you, please, experiment further and find out if returning `127.0.0.1` instead of `NXDOMAIN` fixes the second issue?

**Comment by jackivanov** [comment]
> I'm thinking whether we can prevent the conflicts by using only IPv6 for DNS. It's literally none chances to get a conflict with local subnet at all, especially if we randomise it

Found that it may not work on systems where ipv6 is completely disabled for some reason. Let's stick with randomly generated ipv4 and ipv6 addresses

**Comment by davidemyers** [comment]
@jackivanov `10.19.50.1` seems to work. Wouldn't Algo fail altogether if a client ends up on a local network that uses `10.19.0.0/16`?

@im-batman Maybe we need to give up on trying to make `systemd-resolved` do the right thing and just tell people to install `openresolv`. If you want to test this, be sure to delete any symlink to `resolvconf` you might have previously created.

**Comment by jackivanov** [comment]
@davidemyers I think it wouldn't, unless a client has a resolver running on 10.19.50.1. 
I propose to generate dynamically the last three octets for IPv4, and the first or last three hextets for IPv6

**Comment by jackivanov** [comment]
We can try this out https://github.com/trailofbits/algo/commit/19c3e763669d3a7ad893285a9cf977c685455041
